### DISCUSSION TITLE
Using locally build images when using devcontainer build

### DISCUSSION THREAD

**Discussion by mwoodpatrick** [devcontainers/community#71] [General]
I have a local build of my app which I created and when I run:

`              docker image inspect mwoodpatrick/wp-codespace:latest`

The command reports the expected info.

The image only exists locally and has not been pushed to an external repository. 

but when I run:

`	devcontainer build --workspace-folder .`
	
The command fails since it is looking for:

`         docker.io/mwoodpatrick/wp-codespace:latest`

Is there a way for to tell the command to look for my image locally first?

**Comment by samruddhikhandale** [comment]
Hi 👋 

The cli will first look at the images which are available locally, and only then attempt to pull from the registry. See https://github.com/devcontainers/cli/blob/main/src/spec-node/utils.ts#L216

I tested it, and it did pull the correct local image. @mwoodpatrick Do you mind sharing the logs with the following command, thanks? 👇 
 
```jsonc
devcontainer build --workspace-folder .  --log-level trace
```

**Reply by mwoodpatrick** [reply]
Build log attached
[build.log](https://github.com/devcontainers/community/files/12652370/build.log)

**Comment by samruddhikhandale** [comment]
I wonder if you are facing a similar issue like https://github.com/devcontainers/cli/issues/629

Looking at the logs, looks like you are using `docker-in-docker` Feature, @mwoodpatrick can you use `docker-outside-of-docker` instead? 

Reasoning: `Docker-in-docker` is a separate docker daemon inside of the main container, sort of like nested virtualization. It uses a totally separate /var/lib/docker directory that is not shared with the host. `Docker-outside-of-docker` is a different model however where the host’s docker socket is bind mounted into the container, so docker commands executed from inside the container actually execute on the host.

**Reply by mwoodpatrick** [reply]
Many thanks for the pointer. There are some of my use cases which would be better for using real DIND others will be fine using Docker-outside-of-docker. Anyone have any pointers on best ways to use DIND and sysbox with docker compose. Seems like specifying a local registry for my images would help improve my performance during development rather than always transferring images to docker hub or some other external registry.
